core/schemas/model.py

```diff
@@ -120,46 +125,148 @@ class YetiModel(YetiBaseModel):
+    tag_name = re.sub(r"\s+", "_", tag_name)
+    tag_name = re.sub(r"[^a-zA-Z0-9_.:-]", "", tag_name)
+    return tag_name
 
-    def get_tags(self) -> list[tuple[TagRelationship, "Tag"]]:
-        """Returns the tags linked to this object.
 
-        Returns:
-          A list of tuples (TagRelationship, Tag) representing each tag linked
-          to this object.
-        """
-        from core.schemas.graph import TagRelationship
-        from core.schemas.tag import Tag
+class YetiTagModel(YetiModel):
+    tags: list[YetiTagInstance] = []
+
+    def tag(
+        self,
+        tags: list[str],
+        strict: bool = False,
+        normalized: bool = True,
+        expiration: datetime.timedelta | None = None,
+    ):
+        """Adds or updates a tag in an object."""
+        from core.schemas import tag
+
+        if not isinstance(tags, (list, set, tuple)):
+            raise ValueError("Tags must be of type list, set or tuple.")
+
+        old_tags = [tag.name for tag in self.tags]
+        actual_tags = []
+        for tag_name in tags:
+            new_tag_name = tag_name.strip()
+            if normalized:
+                new_tag_name = normalize_name(new_tag_name)
+            if not new_tag_name:
+                raise RuntimeError(
+                    f"Cannot tag object with empty tag: '{new_tag_name}' -> '{tag_name}'"
+                )
+            actual_tags.append(new_tag_name)
+
+        if strict:
+            self.tags = []
+
+        extra_tags = set()
+        for tag_name in actual_tags:
+            replacements, _ = tag.Tag.filter({"in__replaces": [tag_name]}, count=1)
+
+            if replacements:
+                db_tag = replacements[0]
+            # Attempt to find actual tag
+            else:
+                db_tag = tag.Tag.find(name=tag_name)
+                if not db_tag:
+                    db_tag = tag.Tag(name=tag_name).save()
+
+            action = (
+                message.EventType.new
+                if db_tag.name not in self.tags
+                else message.EventType.update
+            )
+
+            expiration = expiration or db_tag.default_expiration
+            now = datetime.datetime.now(tz=datetime.timezone.utc)
+            for i, tag_instance in enumerate(self.tags):
+                if tag_instance.name == db_tag.name:
+                    self.tags[i].last_seen = now
+                    self.tags[i].expires = now + expiration
+                    self.tags[i].fresh = True
+                    action = message.EventType.update
+                    break
+            else:
+                self.tags.append(
+                    YetiTagInstance(
+                        name=db_tag.name,
+                        last_seen=now,
+                        expires=now + expiration,
+                        fresh=True,
+                    )
+                )
+                action = message.EventType.new
+                db_tag.count += 1
```

Comment:
If using strict, we should `db_tag.count -= 1`. I haven't seen the db_tag.count decrease when unsetting a tag to an object.

Reply:
That is correct; added.